File: backend/src/main/java/com/resulthub/api/workspace/entity/Workspace.java
package com.resulthub.api.workspace.entity;

import com.resulthub.api.user.User;
import com.resulthub.api.workspace.enums.VisibilityMode;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "workspaces")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Workspace {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(nullable = false)
    private String name;

    @Column(unique = true, nullable = false)
    private String slug;

    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @JdbcTypeCode(SqlTypes.NAMED_ENUM)
    private VisibilityMode visibility = VisibilityMode.PUBLIC;

    private String accessCode;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "owner_id")
    private User owner;

    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;
    private LocalDateTime deletedAt;
    private UUID deletedBy;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        if (visibility == null) {
            visibility = VisibilityMode.PUBLIC;
        }
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: backend-mern/src/models/Workspace.ts
import mongoose, { Schema, Document } from 'mongoose';

export enum VisibilityMode {
  PUBLIC = 'PUBLIC',
  PRIVATE = 'PRIVATE',
  RESTRICTED = 'RESTRICTED'
}

export interface IWorkspace extends Document {
  name: string;
  slug: string;
  description?: string;
  category: string;
  visibility: VisibilityMode;
  accessCode?: string;
  owner: mongoose.Types.ObjectId;
  createdAt: Date;
  updatedAt: Date;
  deletedAt?: Date;
  deletedBy?: mongoose.Types.ObjectId;
}

const WorkspaceSchema: Schema = new Schema({
  name: { type: String, required: true },
  slug: { type: String, required: true, unique: true },
  description: { type: String },
  category: { type: String, default: 'Academic', maxlength: 100 },
  visibility: { type: String, enum: Object.values(VisibilityMode), default: VisibilityMode.PUBLIC },
  accessCode: { type: String },
  owner: { type: Schema.Types.ObjectId, ref: 'User', required: true },
  deletedAt: { type: Date },
  deletedBy: { type: Schema.Types.ObjectId, ref: 'User' },
}, {
  timestamps: true
});

export default mongoose.model<IWorkspace>('Workspace', WorkspaceSchema);
